pilot/pkg/serviceregistry/kube/controller/network.go

```diff
@@ -188,24 +188,28 @@ func (c *Controller) extractGatewaysInner(svc *model.Service) bool {
 		}
 	}
 
-	gwsChanged := len(c.networkGateways[svc.Hostname][network]) != len(gws)
-	if !gwsChanged {
-		// number of gateways are the same, check that their contents are the same
-		found := map[model.Gateway]bool{}
-		for _, gw := range gws {
-			found[*gw] = true
-		}
-		for _, gw := range c.networkGateways[svc.Hostname][network] {
-			if _, ok := found[*gw]; !ok {
-				gwsChanged = true
-				break
-			}
-		}
-	}
+	gwsChanged := didGatewaysChange(c.networkGateways[svc.Hostname][network], gws)
 	c.networkGateways[svc.Hostname][network] = gws
 	return gwsChanged
 }
 
+func didGatewaysChange(a, b []*model.Gateway) bool {
+	return len(a) != len(b) || didGatewaysChangeInner(a, b) || didGatewaysChangeInner(b, a)
```

Comment:
Is this too complex, can you refer to `StringSliceEqual`

Reply:
I'm not sure how reliable the ordering will be, and the order shouldn't matter. If we use a similar approach as StringSliceEqual, it could result in unnecessary pushes.

I don't have a strong preference though.

Reply:
Looking through the code it is in a determined order

Reply:
The `StringSliceEqual` is order dependent, but here we don't care about the order.

Reply:
So this is less efficient if the worst-case no change happens.

Reply:
Sure but the number of gateways is usually small. On a no-update it's less efficient here but saves us an eds push